[로또 - 1단계] 정민주 미션 제출합니다

JoungMinJu opened the pull request:
## 고려한 점

1. 로또 생성 전략
기존 로또 생성 전략과 동일하게 적용되도록 구현하였습니다.
- 6자리 숫자 생성
- 중복 숫자 허용 안함
- 각 숫자는 1~45 사이의 숫자

## 고민한 부분
### numberOfLotto와 관련된 고민
로또 개수를 의미하는 numberOfLotto를 Lottos 안에 있는 변수로 관리를 하려다가, 그렇게 되면 List<Lotto> lottos에 값이 더해지거나 빠질 때마다 해당 변수를 같이 관리해줘야한다는 문제가 생길 것이라 판단했습니다!

그리고 numberOfLotto는 "숫자여야한다"는 제약 조건 빼고는 딱히 다른 제약조건이 없었고, 구매금액을 로또 가격으로 나눠서 numberOfLotto를 구하는 로직은 LottoService에서 비즈니스 로직으로 관리해줘도 충분할 것이라 생각하여 해당 로직을 lottoService에 위치시키게 되었습니다아

근데 이렇게 되니 "로또 가격을 LottoService에서 관리하게 되는" 아이러니가 발생하더라구요! 
이 부분은 어떻게 해결해야할지 고민입니다 😮
Lotto 클래스에 price 변수를 넣자니.. numberOfLotto 생성시에 어떻게 가져와야할지도 의문이고........ 리뷰어 분은 어떻게 생각하시는지 의견 여쭙고 싶습니다!!ㅠㅠ

JoungMinJu committed:
add : README 기능 정리

JoungMinJu committed:
add : 로또 한 장(Lotto) 클래스 생성

JoungMinJu committed:
test : 로또 한 장(Lotto) 클래스

JoungMinJu committed:
add : 로또 여러 장(Lottos) 클래스

JoungMinJu committed:
test : 로또 여러 장(Lottos) 클래스

JoungMinJu committed:
add : 로또 여러 장 생성하는 service class

JoungMinJu committed:
test : 로또 여러 장 생성하는 service class

JoungMinJu committed:
add : 로또 숫자 생성기

JoungMinJu committed:
add : 입력 받는 inputView

JoungMinJu committed:
test : 입력 받는 inputView

JoungMinJu committed:
add : 출력하는 outputview

JoungMinJu committed:
add : 로또 controller

JoungMinJu committed:
add : main 함수

JoungMinJu committed:
docs : README 체크

JoungMinJu committed:
refactor : 로또 번호 범위가 1~45가 되도록 수정

fanngineer commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+        for (int count = 0; count < LOTTO_SIZE; count++) {
+            lotto.add(getLottoNumber(lotto));
+        }
+        Collections.shuffle(lotto);
```
Collections.shuffle 메소드가 리스트 내의 요소들을 랜덤으로 섞는 기능인걸로 이해중입니다. 현재 랜덤으로 생성해서 중복체크를 거친 수를 담은 상태인데, shuffle 메소드를 한번 더 적용한 이유를 여쭤보고싶어요!

JoungMinJu replied:
참고해야하는 문서에 shuffle이 있어서 무조건 사용해야하나..?하는 맘에 넣어놓았던 로직입니다 ㅋㅋㅋㅠㅠㅠ 
석환님 코드 리뷰하면서 깨닫게 된(?) 부분인데, 기존 리스트를 shuffle하여 subList 해버리는게 (1) 중복 검사를 할 필요도 없고 (2) 현실에서의 로또 개념과도 더 잘 맞을 것 같습니다 !! 
위와 같은 방식으로 로또 생성하도록 수정하겠습니다 ~! !~ ~

fanngineer commented on src/main/java/service/LottoService.java:
```diff
@@ -0,0 +1,39 @@
+
+    public void printStatusOfLottos(Lottos lottos) {
+        outputView.printNumberOfLotto(lottos.getSize());
+        outputView.printStatusOfLottos(lottos.getStatus());
```
service와 controller로 나누어서 역할 배분하신게 좋은 것 같아요! 그런데 비즈니스 로직에 집중하려면 view에 의존하는 부분은 controller에서 처리하는 것이 좋다고 생각하는데 어떻게 생각하시나요?ㅎㅎ

JoungMinJu replied:
오 좋은 의견 감사합니다 ! ! !  생각해보니 service가 view를 알게 되면 view의 존재 의미가 좀 옅어지는 부분이 있을 것 같아요
참고해서 수정하도록 하겠습니다 ~ ! ~!

fanngineer commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+            lotto.add(getLottoNumber(lotto));
+        }
+        Collections.shuffle(lotto);
+        return lotto;
```
요구사항에는 없지만 저는 예시 output이나 실제 로또에서 숫자가 정렬되어 나오는 점을 참고한점 공유드립니다!

fanngineer commented on src/main/java/service/LottoService.java:
```diff
@@ -0,0 +1,39 @@
+        return new Lottos(numberGenerator, numberOfLotto);
+    }
+
+    private int getNumberOfLotto() {
```
저는 실제 로또를 모티브로해서 LottoMachine이라는 객체에서 로또 생성과 금액 계산을 담당하도록 했습니다. 민주님이 하신 방법대로 필요한 갯수 정보만 넘겨주는 방법도 Lottos객체 관점에서 깔끔한 것 같아요!

JoungMinJu replied:
오 아예 생성하는 객체를 따로 만드셨나봐요 조금 더 좋은 방법인 것 같습니다 ! ! ! 차용해서 구현해보겠습니다  ~~ !~! !

fanngineer commented on src/main/java/service/LottoService.java:
```diff
@@ -0,0 +1,39 @@
+        if (inputPrice <= 0) {
+            return 0;
+        }
+        return inputPrice / LOTTO_PRICE;
```
고민했던 부분인데 2500원 같은 금액이 들어왔을때 나머지를 절삭할까하다 현실에서 1000원 단위의 금액만 가능하게 하는 부분을 생각해서 예외처리한점 공유드릴게요!

JoungMinJu replied:
석환님도 그렇게 구현하셨더라구요!
저는 뭔가 굳이 구입금액이 1,000원 단위일 필요는 없을 것 같다고 생각했어서.. 일단 이 부분은 남겨두겠습니다! 좋은 의견 감사드립니다 :):):)

fanngineer reviewed (commented):
민주님 1단계 하신다고 고생하셨습니다. 제가 개발하면서 생각했던 부분들과 코드 보면서 생각나는 부분들 리뷰로 남겼는데 확인해주시면 감사드리겠습니다!

JoungMinJu committed:
refactor : 코멘트 반영

(1) shuffle 활용한 로또 생성
(2) lottoGenerator 객체 생성
(3) 로또 숫자 정렬
(4) service와 view가 의존하지 않도록 수정

JoungMinJu committed:
test : 코멘트 반영

(1) shuffle 활용한 로또 생성
(2) lottoGenerator 객체 생성
(3) 로또 숫자 정렬
(4) service와 view가 의존하지 않도록 수정

JoungMinJu commented:
(1) shuffle 활용한 로또 생성
(2) lottoGenerator 객체 생성
(3) 로또 숫자 정렬
(4) service와 view가 의존하지 않도록 수정

이렇게 수정한 내용 push 해두었습니다 !! :)

mvc 패턴을 구현할 땐
<img width="389" alt="스크린샷 2024-06-23 오후 1 06 44" src="https://github.com/next-step/java-lotto-clean-playground/assets/81295661/d363203d-eb53-4252-835c-b55d3f01fc7b">
위 그림을 기준으로 구현하였고
service의 경우 model과 controller 사이에 service가 있을 수 있도록 구현하였습니다!! 
(현재 service는 대단한 기능을 하고있진 않은데, 추후 단계를 진행하면서 확장될 가능성이 있을 것 같아 남겨두었습니다 :) 구현하고 나서도 의미가 없다면 삭제할 예정입니다!)
